### PR TITLE
bpo-38307: Add end_lineno attribute to pyclbr _Objects

### PR DESCRIPTION
**Pull request by kebab-mai-haddi**
For back-compatibility, make the new constructor parameter for public classes Function and Class
keyword-only with a default of None.

[bpo-38307](https://bugs.python.org/issue38307): Provide class' and function's end line in order to provide the scope of the classes and functions in a module.

https://bugs.python.org/issue38307

### CODE REVIEW

**Review by terryjreedy — COMMENTED** [review]
Thank you for the revised patch.  On the issue, I questioned whether we can insert end_lineno within the existing signatures.  I have posted to pydev for additional opinions.  The patch otherwise looks fine except for the blurb, which I would revise before merging.

File: Misc/NEWS.d/next/Library/2020-03-16-03-03-21.bpo-38307.2cmw2i.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Adds end line no in class' use: generating dependency.
```
Comment (merwok) [inline Misc/NEWS.d/next/Library/2020-03-16-03-03-21.bpo-38307.2cmw2i.rst]: I can’t parse this line 🙂 
(and second line starting with colon seems strange)

**Comment by terryjreedy** [comment]
Tests/macOS just says 'Error'.  I suspect that this is not related to patch.
Travis and Pipelines failures are due to bad markup in the blurb.

The revised blurb should pass.  It may still be too wordy.

The person merging this should likely add a What's New entry.  The details will depend on where the new argument is added.

File: Misc/NEWS.d/next/Library/2020-03-16-03-03-21.bpo-38307.2cmw2i.rst (hunk 1 above)
Comment (terryjreedy) [reply Misc/NEWS.d/next/Library/2020-03-16-03-03-21.bpo-38307.2cmw2i.rst]: Neither could I nor the doc tests.  I replaced it.

Comment (kebab-mai-haddi) [reply Misc/NEWS.d/next/Library/2020-03-16-03-03-21.bpo-38307.2cmw2i.rst]: Thanks, @terryjreedy !
But how could you push a change to my forked repository?

(I am not objecting to it, I just do not understand how another person can push to mine repo)

**Comment by terryjreedy** [comment]
Since retests had the same 'default role used' complaint, I removed all markup.

File: Misc/NEWS.d/next/Library/2020-03-16-03-03-21.bpo-38307.2cmw2i.rst (hunk 1 above)
Comment (terryjreedy) [reply Misc/NEWS.d/next/Library/2020-03-16-03-03-21.bpo-38307.2cmw2i.rst]: When you made the PR, you left a checkmark in the box that says to allow core developers ('Members of Python' when you mouse over the icons) to modify the files in this branch.  If you had removed it, I would have asked to put is back in if possible.  To pull to your local repository,  which you should do before making further changes, checkout this branch, endline-in-readmodule-module, and, I believe, if you defined 'origin' as the devguide suggests
```
git pull origin endline-in-readmodule-module
```
Don't use rebase.

**Comment by kebab-mai-haddi** [comment]
> Since retests had the same 'default role used' complaint, I removed all markup.

Thanks, @terryjreedy . All tests passed, does this mean the PR will be merged as per the standards?

**Comment by merwok** [comment]
There is a [discussion about this change](https://mail.python.org/archives/list/python-dev@python.org/thread/SQWNNGY7WADHFGAIQZRIMPPLYJGIV4TZ/) : it breaks compatibility but for something that was not supposed to be used directly — more changes are suggested to break with a clear message rather than appear to work.

**Comment by kebab-mai-haddi** [comment]
@terryjreedy , so should I modify the PR with `end_lineno=None`?

File: Lib/pyclbr.py [hunk 2]
```diff
@@ -52,40 +53,41 @@
...
         self.name = name
         self.file = file
         self.lineno = lineno
+        self.end_lineno = end_lineno
         self.parent = parent
         self.children = {}
         if parent is not None:
             parent.children[name] = self
 
 class Function(_Object):
```
Comment (terryjreedy) [inline Lib/pyclbr.py:69]:
```suggestion
# Odd Function and Class signatures are for back-compatibility.
class Function(_Object):
```

File: Lib/pyclbr.py [hunk 3]
```diff
@@ -52,40 +53,41 @@
 
 class _Object:
     "Information about Python class or function."
-    def __init__(self, module, name, file, lineno, parent):
+    def __init__(self, module, name, file, lineno, end_lineno, parent):
```
Comment (terryjreedy) [inline Lib/pyclbr.py:56]: Since `_Object` is private, its signature does not have to be changed.  Neither `end_lineno` nor `parent` are optional here, as they get the defaults from the subclasses.  Arguments for `super_` and `is_async` are not passed on to `_Object`.  So its signature cannot match that of the subclasses.

File: Lib/pyclbr.py [hunk 4]
```diff
@@ -52,40 +53,41 @@
...
         self.parent = parent
         self.children = {}
         if parent is not None:
             parent.children[name] = self
 
 class Function(_Object):
     "Information about a Python function, including methods."
-    def __init__(self, module, name, file, lineno, parent=None, is_async=False):
-        super().__init__(module, name, file, lineno, parent)
+    def __init__(self, module, name, file, lineno, end_lineno, parent=None, is_async=False):
```
Comment (terryjreedy) [inline Lib/pyclbr.py]: Here is the change Guido and I (reluctantly) agreed on.
```suggestion
    def __init__(self, module, name, file, lineno, parent=None, is_async=False, *,  end_lineno=None):
```
Same change for Class.

File: Lib/pyclbr.py [hunk 5]
```diff
@@ -52,40 +53,41 @@
...
         self.super = super_ or []
         self.methods = {}
 
 # These 2 functions are used in these tests
 # Lib/test/test_pyclbr, Lib/idlelib/idle_test/test_browser.py
-def _nest_function(ob, func_name, lineno, is_async=False):
+def _nest_function(ob, func_name, lineno, end_lineno, is_async=False):
     "Return a Function after nesting within ob."
-    return Function(ob.module, func_name, ob.file, lineno, ob, is_async)
+    return Function(ob.module, func_name, ob.file, lineno, end_lineno, ob, is_async)
```
Comment (terryjreedy) [inline Lib/pyclbr.py]:
```suggestion
    return Function(ob.module, func_name, ob.file, lineno, ob, is_async, end_lineno=end_lineno)
```
Same or equivalent change for other 3 Function/Class calls.

**Review by terryjreedy — COMMENTED** [review]
Before merging, an entry for `pyclbr` should be added to Improved Modules in Doc/whatsnew/3.10.rst.  I think the following is enough.

Add an 'end_lineno' attribute to the Class and Function objects that appear in the
tree returned by pyclbr functions.  (Contributed by Aviral Srivastava in bpo-38307.)

(With bpo marked-up as for other entries.)

**Comment by terryjreedy** [comment]
I am making the changes now.

File: Doc/whatsnew/3.10.rst [hunk 6]
```diff
@@ -435,6 +435,13 @@ py_compile
 Added ``--quiet`` option to command-line interface of :mod:`py_compile`.
 (Contributed by Gregory Schevchenko in :issue:`38731`.)
 
+pyclbr
+------
```
Comment (merwok) [inline Doc/whatsnew/3.10.rst:439]:
```suggestion
------

```

Comment (terryjreedy) [reply Doc/whatsnew/3.10.rst:439]: Somehow, a blank line is added to the resulting .html without the blank in the .rst.  But all other entries in the source .rst do have the explicit blank line.  This should too.  Thanks for catching.

Comment (merwok) [reply Doc/whatsnew/3.10.rst:439]: That must mean that docutils is ok to parse the heading here but github isn’t (I suggested this change after seeing bad colorization on the lines that follow)

**Comment by kebab-mai-haddi** [comment]
Thank you @terryjreedy !

Does this mean that new installations of Python will have this change or should one wait for the tag release?
I know making from source will have the change but wasn't sure about which version of Python would have this change. Can you help me understand that?

Thanks again!

**Comment by merwok** [comment]
As a new feature, this will be released in Python 3.10.

**Comment by terryjreedy** [comment]
I don't know what you mean by a tag release, but 3.10.0a5, with Windows and macOS installers should be coming soon and will include this.  3.10.0 should be out next September.
